docs/proposals/20220209-machinepool-machines.md

```diff
@@ -0,0 +1,209 @@
+
+As another example, currently each MachinePool infrastructure provider must implement their own node drain functionality. When MachinePool Machines become available in the CAPI API, all providers can use its existing node drain implementation.
+
+### Goals
+
+- Enable CAPI MachinePool to create Machines to represent infrastructure provider MachinePool Machines
+- Ensure that Machines in MachinePools can participate in the same behaviors as MachineDeployment Machines (node drain, machine health checks, retired node deletion)
+- Enhance `clusterctl describe` to display MachinePool Machines
+
+### Non-Goals/Future Work
+
+- Implement Cluster Autoscaler MachinePool integration (a separate task, although we have a working POC)
+- Allow CAPI MachinePool to include deployment strategies mirroring those of MachineDeployments
+- Move the MachinePools implementation into a `exp/machinepools` package. The current code location makes its separation from `exp/addons` unclear and this would be an improvement.
+- Improve bootstrap token handling for MachinePools. It would be preferable to have a single, short-lived token for each Machine instead of the current shared, long-lived token.
+## Proposal
+
+To create MachinePool Machines, a MachinePool in CAPI needs information about the instances or replicas associated with the provider's implementation of the MachinePool. This information is attached to the provider's MachinePool infrastructure resource in new status fields `InfrastructureMachineSelector` and `InfrastructureMachineKind`. These fields should be populated by the infrastructure provider.
+
+```golang
+// FooMachinePoolStatus defines the observed state of FooMachinePool.
+type FooMachinePoolStatus struct {
+	// InfrastructureMachineSelector is a label query over the infrastructure resources behind MachinePool Machines.
+	// More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#label-selectors
+	// +optional
+	InfrastructureMachineSelector metav1.LabelSelector `json:"infrastructureMachineSelector,omitempty"`
+  // InfrastructureMachineKind is the kind of the infrastructure resources behind MachinePool Machines.
+  // +optional
+  InfrastructureMachineKind string `json:"infrastructureMachineKind,omitempty"`
+}
+```
+
+These fields are an addition to the optional status fields of `InfrastructureMachinePool` in the [provider contract][].
+
+If the fields are populated, CAPI's MachinePool controller will query for the provider-specific infrastructure resources. That query uses the Selector and Kind fields with the API version of the \<Provider\>MachinePool, which is assumed to match the API version of the infrastructure resources.
```

Comment:
I assume the apiVersion assumption holds for the currently known MachinePool implementations? (CAPA, CAPZ, not sure who else)

Reply:
That is my assumption from reading provider code, yes. We could also include an API field explicitly, but I think we can simplify by making this assumption.

Reply:
Makes sense, it's also easy to add an additional backwards compatible field for that if necessary